### PR TITLE
Structure stuff

### LINKED ISSUE #2725: Structure#place has an unnamed flags parameter

### PR DESCRIPTION
**Pull request by liach**
Mostly new structure generation stuff. Some javadocs on hard-to-comprehend areas. Closes #2725.

### CODE REVIEW

**Comment by apple502j** [comment]
I was mapping this exact thing 🤦🏼 See https://github.com/apple502j/yarn/tree/21w39a-structure for my WIP stuff

One thing, I believe `StructurePiecesHolder`'s javadoc ("holder of generated structure pieces") is incorrect now (unless it means "naturally-generated structures"). From what I can tell `class_6626` implements Holder and that class is definitely doing generation stuff (e.g. changing bounding box). `class_6624` is the class that holds already-generated pieces.

**Comment by liach** [comment]
@apple502j ah, that's unfortunate! Feel free to nitpick my pr, pointing out bad names (and feel free to suggest better ones if you have any!)

File: mappings/net/minecraft/structure/PostPlacementProcessor.mapping [hunk 1]
```diff
@@ -0,0 +1,13 @@
+CLASS net/minecraft/class_6621 net/minecraft/structure/PostPlacementProcessor
+	COMMENT A post placment processor for a structure feature runs after all the
```
Comment (enbrain) [inline mappings/net/minecraft/structure/PostPlacementProcessor.mapping]:
```suggestion
	COMMENT A post placement processor for a structure feature runs after all the
```

File: mappings/net/minecraft/structure/StructurePlacementData.mapping [hunk 2]
```diff
@@ -40,5 +41,8 @@ CLASS net/minecraft/class_3492 net/minecraft/structure/StructurePlacementData
 	METHOD method_16444 shouldUpdateNeighbors ()Z
 	METHOD method_16664 removeProcessor (Lnet/minecraft/class_3491;)Lnet/minecraft/class_3492;
 		ARG 1 processor
+	METHOD method_27264 setInitializeMobs (Z)Lnet/minecraft/class_3492;
+		ARG 1 initializeEntities
```
Comment (enbrain) [inline mappings/net/minecraft/structure/StructurePlacementData.mapping]:
```suggestion
		ARG 1 initializeMobs
```

File: mappings/net/minecraft/SharedConstants.mapping [hunk 3]
```diff
@@ -28,7 +28,7 @@ CLASS net/minecraft/class_155 net/minecraft/SharedConstants
 	METHOD method_34872 setGameVersion (Lnet/minecraft/class_6489;)V
 		ARG 0 gameVersion
 	METHOD method_36208 createGameVersion ()V
-	METHOD method_37896 (II)Z
+	METHOD method_37896 shouldSkipFeatureGeneration (II)Z
```
Comment (enbrain) [inline mappings/net/minecraft/SharedConstants.mapping:31]: This actually skips the entire chunk generation. I made a simple mixin to confirm it.
```java
@Mixin(SharedConstants.class)
public class SharedConstantsMixin {
    @Inject(method = "method_37896", at = @At("HEAD"), cancellable = true)
    private static void f(int x, int z, CallbackInfoReturnable<Boolean> cir) {
        cir.setReturnValue(x > 8192 || x < 0 || z > 1024 || z < 0);
    }
}
```
Result:
image

```suggestion
	METHOD method_37896 shouldSkipChunkGeneration (II)Z
```

Comment (liach) [reply mappings/net/minecraft/SharedConstants.mapping:31]: It is like in feature placement chunk status, I recall. Need to check back.

**Comment by liach** [comment]
Will investigate `method_37896` first.

**Comment by enbrain** [comment]
Mojang refactored `method_37896` a bit in 21w44a. It now takes a `ChunkPos` instead of two `int`s.

**Comment by liach** [comment]
Interesting, but unfortunately I am still quite busy recently and this will stay outdated and wip for a while

**Comment by apple502j** [comment]
Updated to 1.18-pre1 and fixed some stuff: https://github.com/apple502j/yarn/tree/pr-2729-structure

**Comment by liach** [comment]
> Updated to 1.18-pre1 and fixed some stuff: https://github.com/apple502j/yarn/tree/pr-2729-structure

feel free to open a pr either to here or me.

**Pull request closed without merging**